### ISSUE TITLE
After upgrade to 2.6.4 - swagger schema uses jsonld for all media types

### ISSUE THREAD

**Issue by guilliamxavier** [api-platform/core#4215]
**API Platform version(s) affected**: 2.6.4 (regression from 2.6.3)

**Description**  
(extracted from #4212)
The generated OpenAPI docs use the same schemas for content mime types `application/json` (and `application/html`) as for `application/ld+json`

**How to reproduce**  
Default config, Dummy resource (id, name)

**Possible Solution**  
After #4213, move both occurrences of
```php
            $schema = new Schema('openapi');
            $schema->setDefinitions($schemas);
```
to inside respectively
`foreach ($responseMimeTypes as $operationFormat) {` and
`foreach ($requestMimeTypes as $operationFormat) {`
(just before each call to `$this->jsonSchemaFactory->buildSchema(/*...*/)`)

**Additional Context**  
<img src="https://user-images.githubusercontent.com/7404452/114721775-8b423200-9d39-11eb-9827-e16e424e08bc.png" width="400" />
(the bug is not in Swagger-UI but really in the generated docs.json)

**Comment by guilliamxavier** [comment]
Now that's interesting... I'm using the default config + `denormalizationContext={"allow_extra_attributes": false}`, and even after both fixes (which I confirm are effective), the generated schemas don't have `"additionalProperties": false`.
But if I customize the config like this:
```yaml
    formats:
        jsonld: ['application/ld+json']
    patch_formats:
        json: ['application/merge-patch+json']
```
then [well, I lose docs.json and Swagger UI, but I can still run the `api:openapi:export` command, and] the generated schemas *do* have `"additionalProperties": false` 🤔
I have also found out that, after coming back to the default config, I can get `"additionalProperties": false` by removing the `$schema->setDefinitions($schemas);` call. Given that it was introduced by a commit titled ["Improve openapi performances"](https://github.com/api-platform/core/pull/4138/commits/e5cf715f8662509e0263cb859e27cb0070215863), maybe it is actually not essential and would be better reverted? I'm not sure of the implications... @soyuka?

**Comment by soyuka** [comment]
Weird as https://github.com/api-platform/core/blob/c81eaae600d9472f65df7d4d8bf41145b06390b8/src/JsonSchema/SchemaFactory.php#L120-L124 should work your way. 

For the patch format to me the only lines modifying something are here: https://github.com/api-platform/core/blob/main/src/Serializer/SerializerContextBuilder.php#L104-L116

From what I know having `$schema->setDefinitions($schemas);` allows to re-use previously generated schemas (as this is really time-consuming for properties). Maybe that the context is lost in the way?

**Comment by guilliamxavier** [comment]
Weird indeed 😔 Speaking of context, https://github.com/api-platform/core/commit/49ce53460b503cd4f77689314766322a57ae5696 replaced `$context` with `null` (and added `, $forceSchemaCollection`) in both calls to `$this->jsonSchemaFactory->buildSchema(/*...*/)` in the body of `private function collectPaths`, making the `array $context` parameter not used anymore; but I have tried reverting these `null` back to `$context` and it doesn't seem to fix anything, worse actually it cancels the effect of removing `$schema->setDefinitions($schemas);` 😕 Anyway, maybe I should open a separate issue?

**Comment by soyuka** [comment]
I wonder though how the schemas differ per-type, maybe that alltogether `jsonld` should not be a part of the schema name?

**Comment by planim** [comment]
I came here with same issue, but i have different solution (Im not sure if this is ok, but it solves also [#4212](https://github.com/api-platform/core/issues/4212)):

I think the real issue is that `\ApiPlatform\Core\OpenApi\Model\Schema` extends `\ArrayObject`, which makes to pass all $schema objects by reference. To fix this i changed in [SchemaFactory.php#L72](https://github.com/api-platform/core/blob/d62e86e9cfa2a7a4eba3029053ff65f43a10e1ce/src/JsonSchema/SchemaFactory.php#L72)
line from:
`$schema = $schema ?? new Schema();`
to
`$schema = clone $schema ?? new Schema();`

Optionally (simpler solution to fix only this ticket):
I think those two lines added in (link: pull request) [OpenApiFactory.php#L258](https://github.com/api-platform/core/blob/472cc39cb8326f87a76f74c55b71fa79b33f3280/src/OpenApi/Factory/OpenApiFactory.php#L258)
`$schema = new Schema('openapi');` and `$schema->setDefinitions($schemas);`

should be moved to both foreach loops [OpenApiFactory.php#L163](https://github.com/api-platform/core/blob/d62e86e9cfa2a7a4eba3029053ff65f43a10e1ce/src/OpenApi/Factory/OpenApiFactory.php#L163) and
[OpenApiFactory.php#L163](https://github.com/api-platform/core/blob/d62e86e9cfa2a7a4eba3029053ff65f43a10e1ce/src/OpenApi/Factory/OpenApiFactory.php#L269)

**Comment by guilliamxavier** [comment]
> I came here with same issue, but i have different solution (Im not sure if this is ok, but it solves also [#4212](https://github.com/api-platform/core/issues/4212)):
> 
> I think the real issue is that `\ApiPlatform\Core\OpenApi\Model\Schema` extends `\ArrayObject`, which makes to pass all $schema objects by reference. To fix this i changed in [SchemaFactory.php#L72](https://github.com/api-platform/core/blob/d62e86e9cfa2a7a4eba3029053ff65f43a10e1ce/src/JsonSchema/SchemaFactory.php#L72)
> line from:
> `$schema = $schema ?? new Schema();`
> to
> `$schema = clone $schema ?? new Schema();`

Mind operator precedence! This is evaluated as
`$schema = (clone $schema) ?? new Schema();`
which is wrong (and will crash when `$schema` is `null`). You must group explicitly
`$schema = clone ($schema ?? new Schema());`
but that will do `$schema = clone new Schema();` (unnecessary `clone`) when `$schema` is `null`. Maybe better would be
`$schema = null !== $schema ? clone $schema : new Schema();`

But that would also change the behavior of `public function buildSchema`, and e.g. https://github.com/api-platform/core/blob/d62e86e9cfa2a7a4eba3029053ff65f43a10e1ce/src/Swagger/Serializer/DocumentationNormalizer.php#L640 would need to be changed (prepend `$schema = `)... On the other hand, it's already used like that elsewhere, notably https://github.com/api-platform/core/blob/d62e86e9cfa2a7a4eba3029053ff65f43a10e1ce/src/Hydra/JsonSchema/SchemaFactory.php#L57 (for jsonld), so maybe that would be for the best after all?

An alternative would be to leave this function as is but change `$schema` to `clone $schema` in *calls* (in both foreach loops inside `private function collectPaths`).

Thanks for the idea 🙂 WDYT @soyuka?

> Optionally (simpler solution to fix only this ticket):
> I think those two lines added in (link: pull request) [OpenApiFactory.php#L258](https://github.com/api-platform/core/blob/472cc39cb8326f87a76f74c55b71fa79b33f3280/src/OpenApi/Factory/OpenApiFactory.php#L258)
> `$schema = new Schema('openapi');` and `$schema->setDefinitions($schemas);`
> 
> should be moved to both foreach loops [OpenApiFactory.php#L163](https://github.com/api-platform/core/blob/d62e86e9cfa2a7a4eba3029053ff65f43a10e1ce/src/OpenApi/Factory/OpenApiFactory.php#L163) and
> [OpenApiFactory.php#L269](https://github.com/api-platform/core/blob/d62e86e9cfa2a7a4eba3029053ff65f43a10e1ce/src/OpenApi/Factory/OpenApiFactory.php#L269)

That's exactly what I suggested under "**Possible Solution**" 😉 (and note that it would fix *not only* this ticket but also #4212, too)

**Comment by planim** [comment]
I agree with You that
`$schema = null !== $schema ? clone $schema : new Schema();`
is best solution and is exactly what i had in mind :)

**Comment by guilliamxavier** [comment]
So I have submitted a PR #4247 to fix the jsonld issue, and opened a separate issue #4248 for additionalProperties
